### PR TITLE
Fixing various regressions

### PR DESCRIPTION
**Pull request by slavapestov**
While testing a user project against the master branch I found a large number of regressions. This PR collects fixes for them. A few I was not able to address but we're tracking them internally.

### CODE REVIEW

File: lib/IRGen/GenCall.cpp [hunk 1]
```diff
@@ -866,10 +866,11 @@ llvm::Type *SignatureExpansion::expandExternalSignatureTypes() {
   // Swift parameters list?
   size_t clangToSwiftParamOffset = paramTys.size();
 
-  // Convert each parameter to a Clang type.
+  // Convert each parameter to a canonical Clang parameter type.
   for (auto param : params) {
     auto clangTy = IGM.getClangType(param);
-    paramTys.push_back(clangTy);
+    auto canClangTy = IGM.getClangASTContext().getCanonicalParamType(clangTy);
+    paramTys.push_back(canClangTy);
```
Comment (rjmccall) [inline lib/IRGen/GenCall.cpp]: I think the better fix here might be to call getUnqualifiedType in ClangTypeConverter::convert's code for getting the type out of a clang::TypeDecl.

Comment (slavapestov) [reply lib/IRGen/GenCall.cpp]: Ok, I'll take a look tomorrow.

**Comment by slavapestov** [comment]
Next reduced test case that needs fixing:

```
import Foundation

extension NSString {
  var x: Float { return 0.0 }
}

var str: String? = nil

func f() {
  _ = str?.x
}
```

And after this, there are 3 more bugs. Sigh...

File: lib/IRGen/GenCall.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/GenCall.cpp]: Ok, that made the change a bit simpler. PTAL

**Comment by slavapestov** [comment]
@CodaFi You might be interested in the parser change for __shared and __owned.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/AST/Module.cpp [hunk 2]
```diff
@@ -34,6 +34,7 @@
 #include "swift/AST/ProtocolConformance.h"
 #include "swift/Basic/Compiler.h"
 #include "swift/Basic/SourceManager.h"
+#include "swift/Parse/Token.h"
```
Comment (jrose-apple) [inline lib/AST/Module.cpp]: How is this better? Just fewer includes?

Comment (slavapestov) [reply lib/AST/Module.cpp]: Yeah, editing Token.h does not trigger recompilation of ~400 source files anymore.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by CodaFi** [comment]
If `__shared` and `__owned` are contextual keywords, there shouldn't be anything stopping `inout` from being given the same treatment.

**Comment by slavapestov** [comment]
@CodaFi inout has legacy behavior where it can appear before a parameter name. I think that's just a hard error now, so if we're OK with losing that, we could make it a contextual keyword. But as Jordan pointed out offline, contextual keywords make error recovery more difficult, so unless we feel there's something to be gained by allowing inout to be used as an identifier, I don't see any reason to do that. I also feel that if we had shared and owned as keywords from the start, they would not need to be contextual, but at this point in time we have source compatibility concerns. That's my only reason for changing this.

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@shahmishal Any ideas what's going on here?

```
 > git --version # timeout=10
 > git fetch --tags --progress git@github.com:apple/swift.git +refs/pull/*:refs/remotes/origin/pr/*
ERROR: Timeout after 10 minutes
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from git@github.com:apple/swift.git
```

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**